pkg/controller/atlasprivateendpoint/atlasprivateendpoint_controller.go

```diff
@@ -0,0 +1,343 @@
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/indexer"
+)
+
+// AtlasPrivateEndpointReconciler reconciles a AtlasPrivateEndpoint object
+type AtlasPrivateEndpointReconciler struct {
+	Scheme                 *runtime.Scheme
+	Client                 client.Client
+	EventRecorder          record.EventRecorder
+	AtlasProvider          atlas.Provider
+	GlobalPredicates       []predicate.Predicate
+	Log                    *zap.SugaredLogger
+	privateEndpointService privateendpoint.PrivateEndpointService
+
+	ObjectDeletionProtection bool
+	independentSyncPeriod    time.Duration
+}
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasprivateendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlasprivateendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasprivateendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlasprivateendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+func (r *AtlasPrivateEndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Infow("-> Starting AtlasPrivateEndpoint reconciliation")
+
+	akoPrivateEndpoint := akov2.AtlasPrivateEndpoint{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &akoPrivateEndpoint, r.Log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), errors.New(result.GetMessage())
+	}
+
+	return r.ensureCustomResource(ctx, &akoPrivateEndpoint)
+}
+
+func (r *AtlasPrivateEndpointReconciler) ensureCustomResource(ctx context.Context, akoPrivateEndpoint *akov2.AtlasPrivateEndpoint) (ctrl.Result, error) {
+	if customresource.ReconciliationShouldBeSkipped(akoPrivateEndpoint) {
+		return r.skip(ctx, akoPrivateEndpoint)
+	}
+
+	conditions := api.InitCondition(akoPrivateEndpoint, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(r.Log, conditions, ctx)
+	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, akoPrivateEndpoint)
+
+	isValid := customresource.ValidateResourceVersion(workflowCtx, akoPrivateEndpoint, r.Log)
+	if !isValid.IsOk() {
+		return r.invalidate(isValid)
+	}
+
+	if !r.AtlasProvider.IsResourceSupported(akoPrivateEndpoint) {
+		return r.unsupport(workflowCtx)
+	}
+
+	var atlasProject *project.Project
+	var err error
+	if akoPrivateEndpoint.Spec.ExternalProject != nil {
+		atlasProject, err = r.getProjectFromAtlas(ctx, akoPrivateEndpoint)
+	} else {
+		atlasProject, err = r.getProjectFromKube(ctx, akoPrivateEndpoint)
+	}
+	if err != nil {
+		return r.terminate(workflowCtx, akoPrivateEndpoint, nil, api.ReadyType, workflow.AtlasAPIAccessNotConfigured, err)
+	}
```

Comment:
Not a blocker for this PR, but I think the approach by @igor-karpukhin on getting the credentials first, then solving the reference is better than trying to do both at once as done here. I would argue we should eventually converge in that direction.